### PR TITLE
*: added pre-commit hook.

### PR DESCRIPTION
**Pull request by zhexuany**
After this, future commited code has to satisfy the rule we specify in
pre-commit.

### CODE REVIEW

**Comment by coocood** [comment]
Good idea!
pre-commit hook is a great tool.
Is this script referenced from other go projects?

**Comment by zhexuany** [comment]
@coocood Yes. It comes from K8s. I will add credits later.

**Comment by ngaut** [comment]
LGTM

**Comment by shenli** [comment]
LGTM
Please add credits or license for the scripts.

File: hooks/pre-commit [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    echo "  gofmt -s -w ${files_need_gofmt[@]}"
+    exit_code=1
+else
+    echo "${green}OK"
+fi
+echo "${reset}"
+
+echo -ne "Checking for files that need goword... "
+files_need_goword=()
+files=($(git diff --cached --name-only --diff-filter ACM | grep "\.go" | grep -v -e "^_vendor"))
```
Comment (hanfei1991) [inline hooks/pre-commit]: What is --diff-filter ACM?

Comment (iamzhout) [inline hooks/pre-commit]: https://git-scm.com/docs/git-diff#git-diff---diff-filterACDMRTUXB82308203

```
--diff-filter=[(A|C|D|M|R|T|U|X|B)…​[*]]
Select only files that are Added (A), Copied (C), Deleted (D), Modified (M), Renamed (R), have their type (i.e. regular file, symlink, submodule, …​) changed (T), are Unmerged (U), are Unknown (X), or have had their pairing Broken (B). Any combination of the filter characters (including none) can be used. When * (All-or-none) is added to the combination, all paths are selected if there is any file that matches other criteria in the comparison; if there is no file that matches other criteria, nothing is selected.

Also, these upper-case letters can be downcased to exclude. E.g. --diff-filter=ad excludes added and deleted paths.
```

Comment (hanfei1991) [inline hooks/pre-commit]: got

**Comment by zhexuany** [comment]
@shenli PTAL

File: Makefile [hunk 2]
```diff
@@ -24,9 +24,9 @@ LDFLAGS += -X "github.com/pingcap/tidb/util/printer.TiDBGitHash=$(shell git rev-
 
 TARGET = ""
 
-.PHONY: all build update parser clean todo test gotest interpreter server dev benchkv benchraw check parserlib
+.PHONY: all build update parser clean todo test gotest interpreter server dev benchkv benchraw check parserlib precommithook
 
-default: server buildsucc
+default: server buildsucc precommithook
```
Comment (coocood) [inline Makefile]: Please do not add it in default.

File: plan/decorrelate.go [hunk 3]
```diff
@@ -23,7 +23,7 @@ import (
 
 // extractCorColumnsBySchema only extracts the correlated columns that match the outer plan's schema.
 // e.g. If the correlated columns from inner plan are [t1.a, t2.a, t3.a] and outer plan's schema is [t2.a, t2.b, t2.c],
-// only [t2.a] is treated as this apply's correlated column.
+// only [t2.a] is treated as this applies correlated column.
```
Comment (coocood) [inline plan/decorrelate.go]: This is unrelated change.

File: Makefile [hunk 4]
```diff
@@ -140,3 +140,7 @@ endif
 	glide vc --only-code --no-tests
 	mkdir -p _vendor
 	mv vendor _vendor/src
+
+precommithook:
```
Comment (coocood) [inline Makefile]: As comments are added in the `pre-commit` file, this is redundant.

Comment (coocood) [inline Makefile]: One time job should not be added to Makefile.

Comment (zhexuany) [inline Makefile]: DONE

File: hooks/pre-commit [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+for file in "${files[@]}"; do
+    # Check for files that fail gofmt.
```
Comment (coocood) [inline hooks/pre-commit]: goword

Comment (zhexuany) [inline hooks/pre-commit]: DONE
